File: easy/logicGates/logicGates.py
import sys

def logic_gate(operation, input1, input2):
    if operation == "AND":
        return "-" if input1 == "-" and input2 == "-" else "_"
    elif operation == "OR":
        return "-" if input1 == "-" or input2 == "-" else "_"
    elif operation == "XOR":
        return "-" if input1 != input2 else "_"
    elif operation == "NAND":
        return "_" if input1 == "-" and input2 == "-" else "-"
    elif operation == "NOR":
        return "_" if input1 == "-" or input2 == "-" else "-"
    elif operation == "NXOR":
        return "_" if input1 != input2 else "-"
    else:
        return None

def main():
    n = int(input())
    m = int(input())

    inputs = {}
    for _ in range(n):
        input_name, input_signal = input().split()
        inputs[input_name] = input_signal

    output_signals = []
    for _ in range(m):
        output_name, gate_type, input_name_1, input_name_2 = input().split()
        input_signal_1 = inputs[input_name_1]
        input_signal_2 = inputs[input_name_2]
        output_signal = "".join(logic_gate(gate_type, input_signal_1[i], input_signal_2[i]) for i in range(len(input_signal_1)))
        inputs[output_name] = output_signal
        output_signals.append((output_name, output_signal))

    for output_name, output_signal in output_signals:
        print(output_name, output_signal)

if __name__ == "__main__":
    main()

File: easy/logicGates/logicGates.c
#include <stdio.h>
#include <string.h>

#define MAX_SIGNAL_LENGTH 100
#define MAX_SIGNALS 16

char logic_gate(char operation, char input1, char input2) {
    switch(operation) {
        case 'A':
            return (input1 == '-' && input2 == '-') ? '-' : '_';
        case 'O':
            return (input1 == '-' || input2 == '-') ? '-' : '_';
        case 'X':
            return (input1 != input2) ? '-' : '_';
        case 'N':
            return (input1 == '-' && input2 == '-') ? '_' : '-';
        default:
            return (input1 == input2) ? '_' : '-';
    }
}

int main() {
    int n, m;
    scanf("%d", &n);
    scanf("%d", &m);
    getchar(); // consume newline character

    char inputs[MAX_SIGNALS][MAX_SIGNAL_LENGTH + 1];
    for (int i = 0; i < n; i++) {
        scanf("%s", inputs[i]);
    }

    for (int i = 0; i < m; i++) {
        char output_name[MAX_SIGNAL_LENGTH + 1], gate_type, input_name_1[MAX_SIGNAL_LENGTH + 1], input_name_2[MAX_SIGNAL_LENGTH + 1];
        char input_signal_1[MAX_SIGNAL_LENGTH + 1], input_signal_2[MAX_SIGNAL_LENGTH + 1], output_signal[MAX_SIGNAL_LENGTH + 1];

        scanf("%s %c %s %s", output_name, &gate_type, input_name_1, input_name_2);

        int index1, index2;
        for (int j = 0; j < n; j++) {
            if (strcmp(input_name_1, inputs[j]) == 0) index1 = j;
            if (strcmp(input_name_2, inputs[j]) == 0) index2 = j;
        }

        strcpy(input_signal_1, inputs[index1]);
        strcpy(input_signal_2, inputs[index2]);

        for (int j = 0; input_signal_1[j] != '\0'; j++) {
            output_signal[j] = logic_gate(gate_type, input_signal_1[j], input_signal_2[j]);
        }
        output_signal[strlen(input_signal_1)] = '\0';
        
        strcpy(inputs[i + n], output_signal);

        printf("%s %s\n", output_name, output_signal);
    }

    return 0;
}